GOES-16/17 Converter - Enhanced MetaData

ericlingerfelt opened the pull request:
## Description

This pull request contains additions to the GOES-16/17 IODAv2 converter and adds the following MetaData variables. 

/MetaData/scan_position
/MetaData/sensor_azimuth_angle
/MetaData/sensor_view_angle
/MetaData/sensor_zenith_angle
/MetaData/solar_azimuth_angle
/MetaData/solar_zenith_angle

This pull request also adds a `reference` directory which contains all documents used to derive formulations and other information used in the converter.

This pull request updates the converter to use a provided resolution for sub-sampling the data from 2km to 256km for smaller file sizes and observation profiles. The new GoesUtil class consolidates the sub-sampling and yaw_flip_flag filter. The GoesLatLon class has been modified to correctly account for longitude coordinates less than -180 degrees. The converter also now removes all nonexistent Lat/Lon values and any data values associated with the bad coordinates. 

### Issue(s) addressed

Link the issues to be closed with this PR
- fixes #615 
- fixes #618 
- fixes #619 
- fixes #624 
- fixes #625

srherbener reviewed (commented):
Thanks for these updates. In general I think the code looks good. I have the following comments:


----------
* The CI test failures are due to a coding norms error. Click on one of the CDash links, then scroll to the bottom of the webpage and click on the link to view test results. The coding norms test should be highlighted (red) for the test failure and you can click on that to see the coding norms error.
---------
* It would be good to add a short unit test at some point. I'm okay with doing that in a subsequent PR.
* The code has a dependency on the solo repository, but it looks like all that is being used out of the solo repo are a couple datetime utilities (Date, JediDate). Would it be possible to use the python datetime module. The reason I'm asking is that it seems a bit heavy-weight to add another dependency to ioda just for a couple datetime utilities. But, if there is something about the solo utilities that is essential (eg, performance) then it's fine to use them.
* If it's okay to switch to the python datetime utilities, then adding a test will be easy and can be done relatively soon.
* If we need to stay with the solo datetime utilities, then we will need to introduce solo into ioda-bundle and include it when we build ioda-converters. Then we can add the test.
----------
* I'm not sure if its good to keep a copy of the references in with the repo.
    * The advantages I see of keeping a copy of the references:
        * They are readily available for people to read
    * The disadvantages I see are: 
        * As you make code changes based on updates with procedures at the data provider, you need to update the references so this introduces sync'ing issues and maintenance tasks
        *  You've already got the URLs for the references listed in the code which might be good enough for people to find and read the references
* It's not really clear to me which is better, what do you think?
----------
* At some point (future PR) it would be good to use the ioda Python API so that you pick up any changes in the ioda file layout automatically through ioda updates.
   * You could use the ioda Python API directly OR use the library that Cory Martin (EMC) started (ioda-converters src/lib-python/ioda_conv_engines.py).

ericlingerfelt commented:
Hi @srherbener, here are my comments -> 

- I will run ctest and correct the coding norms issues. 
- I will remove the Solo dependency. 
- I will remove the documents. The hyperlinks and citations should be enough for other developers. 
- I will create a short unit test in a future PR.
- I will leverage the IODA python API. Probably in the future PR in the previous bullet point. 

Thanks for your help and code review!

ericlingerfelt commented:
Hi @srherbener, I am seeing an issue when I try to build `ioda-converters` either standalone or in with `ioda-bundle`. `ecbuild` returns this at the end of the `Makefile` configuration.

```
CMake Error at /usr/local/share/ecbuild/cmake/ecbuild_add_test.cmake:438 (add_test):
  Error evaluating generator expression:

    $<TARGET_FILE_DIR:_ioda_python>

  No target "_ioda_python"
Call Stack (most recent call first):
  iodaconv/test/CMakeLists.txt:561 (ecbuild_add_test)
```

Any ideas? I was trying to see the coding norms errors for this PR. 

Thanks!

srherbener commented:
> Hi @srherbener, here are my comments ->
> 
> * I will run ctest and correct the coding norms issues.
> * I will remove the Solo dependency.
> * I will remove the documents. The hyperlinks and citations should be enough for other developers.
> * I will create a short unit test in a future PR.
> * I will leverage the IODA python API. Probably in the future PR in the previous bullet point.
> 
> Thanks for your help and code review!

@ericlingerfelt thanks for taking care of these items!

srherbener commented:
> Hi @srherbener, I am seeing an issue when I try to build `ioda-converters` either standalone or in with `ioda-bundle`. `ecbuild` returns this at the end of the `Makefile` configuration.
> 
> ```
> CMake Error at /usr/local/share/ecbuild/cmake/ecbuild_add_test.cmake:438 (add_test):
>   Error evaluating generator expression:
> 
>     $<TARGET_FILE_DIR:_ioda_python>
> 
>   No target "_ioda_python"
> Call Stack (most recent call first):
>   iodaconv/test/CMakeLists.txt:561 (ecbuild_add_test)
> ```
> 
> Any ideas? I was trying to see the coding norms errors for this PR.
> 
> Thanks!

It looks like ioda somehow got built with BUILD_PYTHON_BINDINGS set to OFF. I think at this point, it would work to wipe out the build area and rerun ecbuild using `-DBUILD_PYTHON_BINDINGS=ON`. Then, to be sure, run `make update` followed by `make -j4` (or however many processes you want).

ericlingerfelt commented:
The previous commit took care of the ioda coding norms error. I'm now working on the date issue by removing Solo. The main issue here is that GOES-16/17 data uses day of the year in its time stamp which Solo rectified previously. 

```
Orion-login-2[70] eric$ ctest -R iodaconv_goes_coding_norms
Test project /work/noaa/da/eric/repos/ioda-bundle/build
    Start 404: iodaconv_goes_coding_norms
1/1 Test #404: iodaconv_goes_coding_norms .......   Passed    0.27 sec

100% tests passed, 0 tests failed out of 1

Label Time Summary:
iodaconv    =   0.27 sec*proc (1 test)
script      =   0.27 sec*proc (1 test)

Total Test time (real) =   0.36 sec
```

ericlingerfelt commented:
I have removed the dependency on `solo` and I have conducted another `ioda` coding norms test which passed. I believe that I have addressed all of the comments so far.

BenjaminTJohnson commented on src/goes/goes.py:
```diff
@@ -240,41 +132,29 @@ def _string_to_abisectortype(string):
         log_comp = np.log((self._planck_fk1 / self._rad_data_array) + 1)
-        temp_data_array = ((self._planck_fk2 / log_comp) - self._planck_bc1) / self._planck_bc2
-        self._obsvalue_bt_data_array = np.where(self._dqf_data_array == -999, -999, temp_data_array)
+        self._obsvalue_bt_data_array = ((self._planck_fk2 / log_comp) - self._planck_bc1) / self._planck_bc2
```
Comment: We should probably compare this TB computation with CRTM's internal function for computing TBs, to ensure consistency.    Can you create a comment here identifying the source of this equation/relationship?

ericlingerfelt replied:
The sources for all formulations used in each python class are listed in the header of each class. Yes. I would like to compare this to the TB (assuming that means Temperature Brightness) to the one used in CRTM.

BenjaminTJohnson commented on src/goes/goes_converter.py:
```diff
@@ -174,6 +200,115 @@ def _create_metadata_scan_angle_variables(self):
+        t = start_date.hour + (start_date.minute / 60.0) + (start_date.second / 3600.0)
+        h = -1.0 * ((t - 12.0) / 12.0)
+        j = self._day_of_year
+        declin = -23.45 * np.cos(((2.0 * np.pi * float(j)) / 365.0) + ((20.0 * np.pi) / 365.0))
```
Can you put a comment in this section of the code to identify the source of these relationships (if you know the source)?  This will help with future verification / understanding.

This comment applies to the azimuth section as well.

ericlingerfelt replied:
The sources for all formulations used in each python class are listed in the header documentation of each class.

BenjaminTJohnson commented on src/goes/goes_latlon.py:
```diff
@@ -79,68 +92,126 @@ def _calc_latlon(self):
+        sensor_view_angle = np.nan_to_num(sensor_view_angle, nan=-999)
+        sensor_zenith_angle = np.where(sensor_zenith_angle > 80, 80, sensor_zenith_angle)
+        sensor_zenith_angle = np.where(sensor_zenith_angle < -80, -80, sensor_zenith_angle)
+        sensor_azimuth_angle = np.where(sensor_azimuth_angle > 80, 80, sensor_azimuth_angle)
```
All azimuth angle is not limited to -80 to 80, instead, it goes from 0 to 360.
```
.//src/CRTM_Parameters.f90:  REAL(fp), PUBLIC, PARAMETER :: MAX_SENSOR_AZIMUTH_ANGLE = 360.0_fp
.//src/CRTM_Parameters.f90:  REAL(fp), PUBLIC, PARAMETER :: MAX_SOURCE_AZIMUTH_ANGLE = 360.0_fp
```

ericlingerfelt replied:
OK. I'll change that acceptable range.

BenjaminTJohnson commented on src/goes/goes_latlon.py:
```diff
@@ -79,68 +92,126 @@ def _calc_latlon(self):
+        sqrt_comp = h_sqr + r_eq_sqr - (2.0 * h * r_eq * np.cos(beta))
+        sensor_zenith_angle = np.arcsin((h * np.sin(beta)) / np.sqrt(sqrt_comp)) * 180.0 / np.pi
+        sensor_azimuth_angle = np.arcsin(np.sin(lon_0 - longitude) / np.sin(beta)) * 180.0 / np.pi
+        sensor_view_angle = sensor_zenith_angle
```
Unless I'm missing something, This is incorrect.  

See eq. (1) of the following reference for the conversion:
https://www.researchgate.net/publication/258662879_Comparison_of_two_transmittance_algorithms_in_the_Community_Radiative_Transfer_Model_Application_to_AVHRR

For now, we neglect the index of refraction for air, so the n/n portion is 1.   `H_i` is negligible compared to the height of the satellite, so that's zero.   We may wish to update this in the future, but for now, this will get us close enough.  

The equation would thus be (please verify):  `sensor_view_angle = np.arcsin(np.sin(sensor_zenith_angle) * Rearth/(Rearth+Hsat))`, where Rearth is the nominal radius of earth in km, and Hsat is the altitude of the satellite above the surface in km.    I believe both of these parameters are provided in the metadata for the instrument.

ericlingerfelt replied:
I could not find the correct formulation for sensor view angle and I also saw a few test data examples where they were identical. Thank you for providing the real answer for this formula! I will check out the reference and get back to you.

ericlingerfelt replied:
The equation you reference is for sensor zenith angle and not sensor view angle. The variable theta_sat is labelled as scan angle and not sensor zenith angle. So I am slightly confused. Here are the links I used to calculate these values. https://satepsanone.nesdis.noaa.gov/pub/GASP/documentations/GOESR/GOES-R_SemiStatic_Data_README2.docx and https://www.ncdc.noaa.gov/gridsat/docs/Angle_Calculations.pdf.

BenjaminTJohnson replied:
Maybe I'm confused too.   Alright, let's table this for now.   I recommend creating an issue for this specific topic, because we'll need to ensure that this part is done correctly.

ericlingerfelt replied:
I created this issue to verify these formulations. 

https://github.com/JCSDA-internal/ioda-converters/issues/625

BenjaminTJohnson commented on src/goes/goes_latlon.py:
```diff
@@ -79,68 +92,126 @@ def _calc_latlon(self):
+        sensor_zenith_angle = np.where(sensor_zenith_angle > 80, 80, sensor_zenith_angle)
+        sensor_zenith_angle = np.where(sensor_zenith_angle < -80, -80, sensor_zenith_angle)
+        sensor_azimuth_angle = np.where(sensor_azimuth_angle > 80, 80, sensor_azimuth_angle)
+        sensor_azimuth_angle = np.where(sensor_azimuth_angle < -80, -80, sensor_azimuth_angle)
```
Same azimuth angle comment.

ericlingerfelt replied:
OK. I'll change that acceptable range.

BenjaminTJohnson commented on src/goes/goes_util.py:
```diff
@@ -0,0 +1,117 @@
+        Sets the fill_value_index_array
+        fill_value_index_array - an array of indices to remove
+        """
+        self._fill_value_index_array = fill_value_index_array
```
Comment: probably good to verify the usage of fill value here, and that it's consistent with code expectations if you haven't done so already.

ericlingerfelt replied:
I will add more comments in the code regarding this.

BenjaminTJohnson commented:
@ericlingerfelt I've added my review.  Please feel free to ask any questions for clarification etc., I will respond ASAP.

ericlingerfelt commented on src/goes/goes.py:
```diff
@@ -50,22 +51,21 @@ def _get_metadata_from_input_file_path(self):
-        self._metadata_dict['end_date'] = Date(metadata_array[4][1:-1])
-        self._metadata_dict['creation_date'] = Date(metadata_array[5][1:-1])
+        year = int(metadata_array[3][1:-1][0:4])
+        day_of_year = int(metadata_array[3][1:-1][4:7])
```
Yes. You are right. The second array index is exclusive.

ericlingerfelt commented:
> @ericlingerfelt I've added my review. Please feel free to ask any questions for clarification etc., I will respond ASAP.

Thanks, @BenjaminTJohnson!

gthompsnJCSDA commented on src/goes/test_goes_converter.py:
```diff
@@ -30,6 +35,8 @@ def test_goes_converter(input_file_paths, latlon_file_path, output_file_path_rf,
     output_file_path_bt = sys.argv[4]
-    test_goes_converter(input_file_paths, latlon_file_path, output_file_path_rf, output_file_path_bt)
+    resolution = sys.argv[5]
+    resolution = sys.argv[6]
```
It appears that you have an extra assignment statement here and I can only assume that you need only one of these two - not sure which one.

ericlingerfelt replied:
This has been resolved.

gthompsnJCSDA commented on src/goes/goes.py:
```diff
@@ -240,41 +132,29 @@ def _string_to_abisectortype(string):
         if 'M2' in string:
```
Are we really planning to handle incoming mesoscale region data?  The lat/lon values would need to be computed nearly daily for real-time data as the NWS picks regions of weather to suit the needs of operational forecasting.  Since I know that we are creating static lat/lon files intended to be read from disk to save time, then I hope the naming of the lat/lon static files has some identifier for full disk and conus at minimum since these should remain very nearly static always.

ericlingerfelt replied:
This will be addressed in a future PR. Currently, we are only using full disk scans for the Data Lake project. Please remind me of this later on.

gthompsnJCSDA commented on src/goes/goes_converter.py:
```diff
@@ -72,27 +81,44 @@ def _check_arguments(self):
-    def _create_input_data_file_dicts(self):
+    def _initialize(self):
         """
         Create two local dictionaries contained the Goes class instances for brightness temperature (ABI channels 1-6)
```
The comment here is mistaken.  Channels 1-6 are reflectances whereas channels 7-16 are the infrared brightness temperatures.

ericlingerfelt replied:
This has been resolved.

gthompsnJCSDA commented on src/goes/goes_converter.py:
```diff
@@ -117,152 +143,266 @@ def _get_nadir_attribute_template(self):
-        self._output_dataset_bt.variables['nlocs'][:] = np.arange(1, nlocs + 1, 1, dtype='int32')
+        output_dataset.createDimension('nlocs', nlocs)
+        output_dataset.createVariable('nlocs', 'i4', 'nlocs')
+        output_dataset.variables['nlocs'].setncattr('suggested_chunk_dim', nlocs)
```
I know the chunk_dim is used for the compression algorithm, but shouldn't it be set lower than the number of total locations?  Isn't it more optimal to chunk the vector is smaller units that compress smaller by using semi-local min/max/range info?

ericlingerfelt replied:
This is a placeholder until an optimized chuck dimension has been selected. This will be addressed in a future PR.

gthompsnJCSDA commented on src/goes/goes_converter.py:
```diff
@@ -308,18 +448,19 @@ def _create_preqc_brightness_temperature_variable(self):
+        output_dataset['/PreQC/brightness_temperature'].setncattr('flag_meanings',
+                                                                  'good_pixel_qf '
+                                                                  'conditionally_usable_pixel_qf '
+                                                                  'out_of_range_pixel_qf no_value_pixel_qf')
```
Is this flag meanings section done correctly?  If I was to guess, I would expect flag value 0 to mean 'good_pixel', flag value 1 to mean 'conditionally_usable_pixel', flag value 2 to mean 'out_of_range_pixel' and flag value 3 to mean 'no_value_pixel.'  The way it's written (here plus a duplicate in next method) appears to have a space between the out-of-range and no-value so I would be expecting the list to be 4 items for flag_meanings.

ericlingerfelt replied:
This has been resolved in the latest commit.

gthompsnJCSDA reviewed (commented):
I sprinkled in comments where I think attention may be needed.

It would be tremendously helpful to visualize what the code does by adding some plots.  For example, if you start with the 2km common remapped (all 16 channels) and make a plot of one or two of them, and then make another plot with the re-sampled from 2km to 4km, then up to 8km.  No need to go super big, but just some visualizations to see the code is doing as advertised.  Similarly, it might be good to see plots of the final output of some of the angles to ensure that all appears to work.

srherbener commented:
@gthompsnJCSDA are you okay with merging this PR (once the CI test failures are addressed). Unless the issues you brought up are show-stoppers, my preference would be to get this merged now and address outstanding issues in subsequent PRs. Thanks!

ericlingerfelt commented:
Please wait on the merge. @fabiolrdiniz and I almost have all of the angles corrected with images. I will be committing to this branch later today. Thanks for your patience!

srherbener commented:
> Please wait on the merge. @fabiolrdiniz and I almost have all of the angles corrected with images. I will be committing to this branch later today. Thanks for your patience!

Thanks for the update @ericlingerfelt. I'll wait for the commit before merging.

ericlingerfelt commented:
@srherbener, please merge at your leisure.

srherbener commented:
@ericlingerfelt we are still getting the coding norms errors:
```
Test output
+ SRC_DIR=/jcsda/ioda-bundle/iodaconv/src/goes
+ PYCODESTYLE_CFG_PATH=/jcsda/ioda-bundle/iodaconv
+ cd /jcsda/ioda-bundle/iodaconv/src/goes
+ pycodestyle -v --config=/jcsda/ioda-bundle/iodaconv/.pycodestyle --filename=*.py,*.py.in .
cli configuration: /jcsda/ioda-bundle/iodaconv/.pycodestyle
directory .
checking ./goes.py
checking ./goes_converter.py
./goes_converter.py:257:29: W503 line break before binary operator
./goes_converter.py:258:29: W503 line break before binary operator
./goes_converter.py:264:19: W503 line break before binary operator
./goes_converter.py:265:19: W503 line break before binary operator
./goes_converter.py:268:23: W503 line break before binary operator
checking ./goes_latlon.py
checking ./goes_util.py
checking ./test_goes_converter.py
5       W503 line break before binary operator
```

These are a pain, but there appears to be two ways to fix this (https://www.python.org/dev/peps/pep-0008/#maximum-line-length)
* use line continuation (backslash at the end of the line)
* use parantheses (ie, around the entire expression on the lhs of the assignment)

Here's the first one the checker is complaining about:
https://github.com/JCSDA-internal/ioda-converters/blob/f03cd4e5be0524bed03eac4a125c79228a7c2bae/src/goes/goes_converter.py#L256-L258

So you should be able to fix it by placing another set of `()` around the entire expression.

ericlingerfelt commented:
Hi @srherbener, I'll take care of this right now.

ericlingerfelt commented:
I don't think the extra parenthesis will do it. I have them around the expression at line 264.

srherbener commented:
How about the line continuation (backslash at the end of the line)?

ericlingerfelt commented:
Hi @srherbener, the latest commit removes the issue by breaking up the equations into components.

srherbener commented:
@ericlingerfelt thank you for your patience with the coding norms errors. That seems like a pain to have to break up the computations like you did. 

I'll wait for the tests to finish and merge if all looks okay.
